File: cpp/Binary_Tree_Level_Order_Traversal.cc
/*
Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

For example:
Given binary tree {3,9,20,#,#,15,7},
    3
   / \
  9  20
    /  \
   15   7
return its level order traversal as:
[
  [3],
  [9,20],
  [15,7]
]
confused what "{1,#,2,3}" means? > read more on how binary tree is serialized on OJ.
*/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution 
{
public:
    vector<vector<int> > levelOrder(TreeNode *root) 
    {
        vector<vector<int> > result;
        if (root == NULL)
            return result;
            
        queue<TreeNode*> current, next;
        vector<int> level; 
        
        current.push(root);
        while (!current.empty())
        {
            while (!current.empty())
            {
                TreeNode* node = current.front();
                current.pop();
                
                level.push_back(node->val); 
                
                if (node->left != NULL)
                    next.push(node->left); 
                    
                if (node->right != NULL)
                    next.push(node->right);
            }
            
            result.push_back(level); 
            level.clear();
            swap(next, current); 
        }
    }
};

File: go/binary-tree-level-order-traversal/binary-tree-level-order-traversal.go
package binary_tree_level_order_traversal

import (
	"../util"
)

/*
给定一个二叉树，返回其按层次遍历的节点值。 （即逐层地，从左到右访问所有节点）。

例如:
给定二叉树: [3,9,20,null,null,15,7],

    3
   / \
  9  20
    /  \
   15   7
返回其层次遍历结果：

[
  [3],
  [9,20],
  [15,7]
]
*/

type TreeNode = util.TreeNode

func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	result := make([][]int, 0)
	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	for len(queue) > 0 {
		level := make([]int, 0)

		l := len(queue)
		for i := 0; i < l; i++ {
			cur := queue[0]
			queue = queue[1:]
			level = append(level, cur.Val)

			if cur.Left != nil {
				queue = append(queue, cur.Left)
			}
			if cur.Right != nil {
				queue = append(queue, cur.Right)
			}
		}
		result = append(result, level)
	}

	return result
}
